src/main/java/com/powsybl/balances_adjustment/util/ControlArea.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.balances_adjustment.util;
+
+import java.util.*;
+
+import com.powsybl.cgmes.extensions.CgmesControlAreas;
+import com.powsybl.iidm.network.*;
+import com.powsybl.math.graph.TraverseResult;
+import com.powsybl.math.graph.UndirectedGraph;
+import com.powsybl.math.graph.UndirectedGraphImpl;
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public class ControlArea implements NetworkArea {
+
+    private final Set<Object> terminalsAndBoundaries = new HashSet<>();
+    private final Set<String> voltageLevelIdsCache;
+
+    public ControlArea(Network network, String controlAreaId) {
+        CgmesControlAreas cgmesControlAreaMapping = network.getExtension(CgmesControlAreas.class);
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getTerminals());
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getBoundaries());
+
+        voltageLevelIdsCache = getVoltageLevelIds(network, terminalsAndBoundaries);
+    }
+
+    @Override
+    public double getNetPosition() {
+        return terminalsAndBoundaries.parallelStream().mapToDouble(o -> {
+            if (o instanceof Terminal) {
+                return getLeavingFlow((Terminal) o);
+            } else if (o instanceof Boundary) {
+                return -((Boundary) o).getP();
+            }
+            throw new AssertionError();
+        }).sum();
+    }
+
+    @Override
+    public Collection<String> getVoltageLevelIds() {
+        return Collections.unmodifiableSet(voltageLevelIdsCache);
+    }
+
+    private static Set<String> getVoltageLevelIds(Network network, Set<Object> terminalsAndBoundaries) {
+        Set<String> voltageLevelIds = new HashSet<>();
+        Map<VoltageLevel, Integer> vlNums = new HashMap<>();
+        UndirectedGraph<Object, Object> voltageLevelGraph = createVoltageLevelGraph(network, vlNums);
+        for (Object o : terminalsAndBoundaries) {
+            voltageLevelGraph.traverse(vlNums.get(getVoltageLevel(o)), (v1, e, v2) -> findContainedVoltageLevels(voltageLevelGraph.getEdgeObject(e),
```

Comment:
I'm not sure to understand what you have in mind. You mean that I could start by the voltage levels in the network ? I'm not sure it will work if we start in a voltage level outside the area.

Reply:
Or do you mean that we could use a condition such as 
```java
            if (voltageLevelIds.contains(getVoltageLevel(o).getId())) {
                continue;
            }
```
?

Reply:
Yes, exactly, sorry for not being clear!